### ISSUE TITLE
subscriptions in phoenix 14

### ISSUE THREAD

**Issue by cultofmetatron** [absinthe-graphql/absinthe_phoenix#57]
Hi, currently trying to setup subscriptions in phoenix 1.4 and currently having difficulty getting it setup

I setup a subcription

  ```elixir
object :order_cycle_subscriptions do
    field :order_cycle_created, :order_cycle do
      arg :restaurant_id, :id |> non_null()
      config fn args, context ->
        IO.puts("=====>>    configured topi")
        IO.inspect(args)
        IO.inspect(context)
        {:ok, topic: args.restaurant_id}
      end
      trigger :create_order_cycle, topic: fn order_cycle ->
        IO.puts("=====>>    triggerring create order cyle")
        IO.inspect(order_cycle)
        order_cycle.restaurant_id
      end
    end
  end
```

now when I run the  :create_order_cycle, I can confirm that the trigger callback runs but the config callback does not. at what point in the lifecycle does that config run? I’m currently not getting any values when I subscribe to this channel from graphiql but I do see the subscriptoon listen in the graphiql schema viewer (edited) 

in my router I have this set
  ```elixir
scope "/api/v1/gql" do
    pipe_through :api
    forward "/graphiql", Absinthe.Plug.GraphiQL,
      schema: BlinqWeb.Schema,
      socket: BlinqWeb.GqlSubscriptionSocket,
      interface: :playground,
      json_codec: Jason
    forward "/", Absinthe.Plug,
      schema: BlinqWeb.Schema,
      socket: BlinqWeb.GqlSubscriptionSocket,
      json_codec: Jason
  end
```

currently ot even trying to enforce authentication so
```elixir
defmodule BlinqWeb.GqlSubscriptionSocket do
  use Phoenix.Socket
  use Absinthe.Phoenix.Socket,
    schema: BlinqWeb.Schema

  #transport :websocket, Phoenix.Transports.WebSocket

  def connect(params, socket) do
    #current_user = current_user(params)
    #socket = Absinthe.Phoenix.Socket.put_opts(socket, context: %{
    #  current_user: current_user
    #})
    {:ok, socket}
  end

  defp current_user(%{"user_id" => id}) do
    #MyApp.Repo.get(User, id)
  end

  def id(_socket), do: nil


end
```

I’m using phoenix 1.4 with absinthe 1.4.16

in the console I see a 400 hit `/api/v1/gql/graphiql/websocket` on subscription from graphiql.

I did make some minor deviations based on https://github.com/absinthe-graphql/absinthe_phoenix/issues/52. in addition I added 

```elixir
  socket "/api/v1/gql/", BlinqWeb.GqlSubscriptionSocket,
    websocket: true,
    longpoll: false
```

to my Endpoint file and eschewed the `transport :websocket, Phoenix.Transports.WebSocket` line in my socket file as 1.4 deprecates it and I have a feeling that might have something to do with the websocket stuff not working

I'm getting no other errors popping up when I start up the server

**Comment by lexun** [comment]
I believe this is happening because your subscription is not on the root level subscription object. I'm not sure if that's just the behavior of this library or part of the spec, but I tested locally and had the same results.

**Comment by cultofmetatron** [comment]
@lexun  thatnks for lookinginto it! I moved the subscription to the root object in my schema

```elixir

  subscription do
    #import_fields :order_cycle_subscriptions

    field :order_cycle_created, :order_cycle do
      arg :restaurant_id, :id |> non_null()
      config fn args, context ->
        IO.puts("=====>>    configured topi")
        IO.inspect(args)
        IO.inspect(context)
        {:ok, topic: args.restaurant_id}
      end
      trigger :create_order_cycle, topic: fn order_cycle ->
        IO.puts("=====>>    triggerring create order cyle")
        IO.inspect(order_cycle)
        order_cycle.restaurant_id
      end
    end
  end

```

same issue. perhaps I misunderstood what you meant?


to reiterate, that `IO.puts("=====>>    triggerring create order cyle")` executes but not the config callback and I can't connect to the websocket  via graphiql

**Comment by lexun** [comment]
What are you using for "WS URL" in GraphiQL?

**Comment by cultofmetatron** [comment]
TLDR: ws://localhost:4000/api/v1/gql/graphiql

```elixir
  scope "/api/v1/gql" do
    pipe_through :api
    forward "/graphiql", Absinthe.Plug.GraphiQL,
      schema: BlinqWeb.Schema,
      socket: BlinqWeb.GqlSubscriptionSocket,
      interface: :playground,
      json_codec: Jason,
      socket_url: "ws://localhost:4000/api/v1/gql/graphiql"
    forward "/", Absinthe.Plug,
      schema: BlinqWeb.Schema,
      socket: BlinqWeb.GqlSubscriptionSocket,
      json_codec: Jason
  end
```

graphiql tries to access GET /api/v1/gql/graphiql/websocket and gets a 400

**Comment by lexun** [comment]
Could you try changing it to `ws://localhost:4000/api/v1/gql`?
And drop `socket_url` from the config?

**Comment by benwilson512** [comment]
> at what point in the lifecycle does that config run? 

Config is run when any subscription document is submitted. It is only run at that point, it is not necessary for it to run when a subscription is triggered. Can you IO.inspect the return value of your topic and trigger functions?

One issue you may have is that you have a socket path that is identical to a router path. You can't serve websockets and HTTP requests from the same path.

In fact you shouldn't even need to specify the socket URL if you have the `socket` specified, it will infer the URL from the socket name. I think your main issue is the URL collision.

**Comment by cultofmetatron** [comment]
@benwilson512 @lexun Thankyou very much for the help!!

I removed the /graphql

```
  scope "/api/v1/gql" do
    pipe_through :api
    forward "/", Absinthe.Plug.GraphiQL,
      schema: BlinqWeb.Schema,
      socket: BlinqWeb.GqlSubscriptionSocket,
      interface: :playground,
      json_codec: Jason
      #socket_url: "ws://localhost:4000/api/v1/gql/graphiql"
    forward "/", Absinthe.Plug,
      schema: BlinqWeb.Schema,
      socket: BlinqWeb.GqlSubscriptionSocket,
      json_codec: Jason
  end

```

and it looks like subscriptions are working!
